### PR TITLE
Add support for containers of move only typed values to range.

### PR DESCRIPTION
**Pull request by thomasspriggs**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: unit/util/range.cpp [hunk 1]
```diff
@@ -65,3 +65,95 @@ SCENARIO("range tests", "[core][util][range]")
...
+      REQUIRE(both_inputs[13].value == 4);
+      REQUIRE(both_inputs[14].value == 5);
+      REQUIRE(both_inputs[15].value == 6);
+      REQUIRE(both_inputs[16].value == 7);
+      REQUIRE(both_inputs[17].value == 8);
+      REQUIRE(both_inputs[18].value == 9);
+      REQUIRE(both_inputs[19].value == 10);
+    }
+  }
+}
```
Comment (romainbrenguier) [inline unit/util/range.cpp]: :pick: missing new line

File: unit/util/range.cpp [hunk 2]
```diff
@@ -65,3 +65,95 @@ SCENARIO("range tests", "[core][util][range]")
...
+  "[core][util][range]")
+{
+  GIVEN("A vector of move only typed values.")
+  {
+    std::vector<move_onlyt> input1;
+    for(int i = 1; i <= 10; ++i)
+      input1.emplace_back(i);
+    std::vector<move_onlyt> input2;
+    for(int i = 1; i <= 10; ++i)
+      input2.emplace_back(i);
```
Comment (romainbrenguier) [inline unit/util/range.cpp]: :pick: input2 is not used in all the tests, consider moving it closer to the test actually using it (maybe in a second GIVEN)

**Review by romainbrenguier — APPROVED** [review]

File: unit/util/range.cpp [hunk 3]
```diff
@@ -65,3 +65,95 @@ SCENARIO("range tests", "[core][util][range]")
...
+
+class move_onlyt
+{
+public:
+  move_onlyt(move_onlyt &&) = default;
+  move_onlyt &operator=(move_onlyt &&) = default;
+  move_onlyt(const move_onlyt &) = delete;
+  move_onlyt &operator=(const move_onlyt &) = delete;
+
+  explicit move_onlyt(int value) : value{value} {};
```
Comment (thk123) [inline unit/util/range.cpp:77]: :pick: I'd have probably done an actually move-only type (e.g. with a `std::unique_ptr`) so I'd be sure that I'd really fixed it.

File: unit/util/range.cpp (hunk 1 above)
Comment (thk123) [inline unit/util/range.cpp]: :no_entry_sign: missing new line

File: src/util/range.h [hunk 4]
```diff
@@ -142,6 +142,16 @@ class filter_iteratort
     return tmp;
   }
 
+  value_type &operator*()
```
Comment (thk123) [inline src/util/range.h:146]: :question: The commit message + change seem unrelated to me, perhaps you could expand in the commit message why these are required.

File: src/util/range.h [hunk 5]
```diff
@@ -244,6 +244,20 @@ struct concat_iteratort
...
```
Comment (thk123) [inline src/util/range.h:248]: :question: Ditto - please elaborate.

**Review by thk123 — COMMENTED** [review]
Could the commit messages contain a little more info why these things are required.

File: unit/util/range.cpp (hunk 3 above)
Comment (thomasspriggs) [reply unit/util/range.cpp:77]: 'move_onlyt' is an actual move only type, because I have deleted to copy constructor and the copy assignment operator. Attempting to copy it will result in a compile error.

File: src/util/range.h [hunk 6]
```diff
@@ -357,12 +357,11 @@ ranget<iteratort> make_range(iteratort begin, iteratort end)
   return ranget<iteratort>(begin, end);
 }
```
Comment (tautschnig) [inline src/util/range.h:388]: Comment on the commit message: s/in oder/in order/

**Review by tautschnig — APPROVED** [review]
Some minor bits left as noted by reviewers.

**Comment by thomasspriggs** [comment]
I have addressed the review comments. I have also added better move support and corresponding unit tests for the `map` operation, as I realised I had previously missed this.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: cdcd544d582aa877ffc43948315be5c4b647f990).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95016060
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by thomasspriggs** [comment]
I have rebased in order to re-check test-gen compatibility.

File: unit/util/range.cpp [hunk 7]
```diff
@@ -65,3 +65,121 @@ SCENARIO("range tests", "[core][util][range]")
...
+      REQUIRE(both_inputs[3].value == 7);
+      REQUIRE(both_inputs[4].value == 8);
+      REQUIRE(both_inputs[5].value == 9);
+    }
+  }
+  GIVEN("A const vector of ints.")
+  {
+    const std::vector<int> input{1, 2, 3, 4, 5};
+
+    THEN("The vector can be mapped into a range of move-only types")
```
Comment (thk123) [inline unit/util/range.cpp:170]: If it works - you should add a test that maps from a move only type to a regular type, and one that moves the same type through the map

**Review by thk123 — COMMENTED** [review]
New changes look good and previous comments addressed - if possible I'd like to see two more tests for `map`

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0b73746794271d3276a8ba0df02484f3d5657756).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95095292

**Comment by thk123** [comment]
Also @romainbrenguier should definitely re-review the unique_ptr -> shared_ptr before merging.

File: unit/util/range.cpp (hunk 7 above)
Comment (thomasspriggs) [reply unit/util/range.cpp:170]: The test which "maps from a move only type to a regular type" already exists. As we discussed, you missed it when reviewing. The `f` passed into `map` can't move from its parameter, because the parameter has to be a const reference. I have expanded the doxygen to document the extent of the support for move-only types and its limitations.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bbc201668e32867361841f2ad4eba3ecafe3185b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95102012

**Review by romainbrenguier — APPROVED** [review]
The change from unique to shared pointer looks good
